### PR TITLE
Move StringSegment to primitives

### PR DESCRIPTION
**Pull request by rynowak**
Working on adding some tests....

### CODE REVIEW

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace Microsoft.Extensions.Primitives
+{
+    public struct StringSegment : IEquatable<StringSegment>
+    {
+        private readonly string _buffer;
+        private readonly int _offset;
+        private readonly int _count;
+
+        public StringSegment(string buffer, int offset, int count)
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: /cc @Tratcher @davidfowl - this constructor is used in HttpAbstractions with `offset = -1` to signify an invalid/null segment - this is a bit greasy to put on a public API.

Suggestion: Validate parameters here, require `buffer != null` and valid `offset` and `count` - use the default constructor for `invalid`.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        public int Count
+        {
+            get { return _count; }
+        }
+
+        public string Value
+        {
+            get { return _offset == -1 ? null : _buffer.Substring(_offset, _count); }
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: These are fragile, and need to handle the case where the default constructor is used.

**Comment by Tratcher** [comment]
Where will this be used?

**Comment by rynowak** [comment]
Routing

**Comment by rynowak** [comment]
https://github.com/aspnet/Routing/pull/215

**Comment by rynowak** [comment]
This is another one of those things like `StringValues` that's way more generate purpose than just our usage :laughing:

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+using System;
+
+namespace Microsoft.Extensions.Primitives
+{
+    public struct StringSegment : IEquatable<StringSegment>
+    {
+        private readonly string _buffer;
+        private readonly int _offset;
+        private readonly int _count;
```
Comment (dougbu) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Could you get rid of the backing fields as part of this move? Get-only auto-properties seem to work fine in `struct`s now.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                int hashCode = (_buffer != null ? _buffer.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ _offset;
+                hashCode = (hashCode * 397) ^ _count;
+                return hashCode;
```
Comment (dougbu) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: use `HashCodeCombiner`

**Comment by dougbu** [comment]
:watch: for tests

**Comment by rynowak** [comment]
@Tratcher updated - do you have any concerns about adding parameter validation here?

File: Common.sln [hunk 5]
```diff
@@ -27,8 +26,6 @@ Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Extensions.Securi
 EndProject
 Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Extensions.HashCodeCombiner.Sources", "src\Microsoft.Extensions.HashCodeCombiner.Sources\Microsoft.Extensions.HashCodeCombiner.Sources.xproj", "{31D4C34C-436F-48E9-B8E4-641574741887}"
 EndProject
-Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "Microsoft.Extensions.BufferedHtmlContent.Sources", "src\Microsoft.Extensions.BufferedHtmlContent.Sources\Microsoft.Extensions.BufferedHtmlContent.Sources.xproj", "{B1B2B906-24AE-4C57-AAC5-19B734014504}"
```
Comment (rynowak) [inline Common.sln]: looks like someone forgot the update this

File: NuGet.config [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (rynowak) [inline NuGet.config]: Not going to fight vs on this one. It insists on there being a BOM here.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        {
+            get { return HasValue ? Buffer.Substring(Offset, Count) : null; }
+        }
+
+        public bool HasValue
+        {
+            get { return Buffer != null; }
+        }
+
+        public bool Equals(StringSegment other)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Change Equals and GetHashCode to only look at the in-range data. Also add an overload of Equals(StringSegment, StringComparison)

**Comment by rynowak** [comment]
Updated

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+                return false;
+            }
+
+            return string.Compare(Buffer, Offset, text, 0, textLength, comparisonType) == 0;
+        }
+
+        public override int GetHashCode()
+        {
+            var hash = HashCodeCombiner.Start();
+            hash.Add(Value);
```
Comment (rynowak) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: This means that `GetHashCode` will allocate. If this becomes an issue anywhere we need it, we can add caching, or a custom implementation to hash the substring.

**Comment by Tratcher** [comment]
:shipit:

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**